packages/gatsby/src/query/queue.js

```diff
@@ -67,35 +70,43 @@ const createDevelopQueue = getRunner => {
  * Note: queue is reused in develop so make sure to thoroughly cleanup hooks
  */
 const processBatch = async (queue, jobs, activity) => {
-  let numJobs = jobs.length
+  const numJobs = jobs.length
   if (numJobs === 0) {
     return Promise.resolve()
   }
 
   return new Promise((resolve, reject) => {
+    let taskFinishCallback
     if (activity.tick) {
-      queue.on(`task_finish`, () => activity.tick())
+      taskFinishCallback = () => activity.tick()
+      queue.on(`task_finish`, taskFinishCallback)
+    }
+
+    const taskFailedCallback = (...err) => {
+      gc()
+      reject(err)
+    }
+
+    const drainCallback = () => {
+      gc()
+      resolve()
     }
 
     const gc = () => {
-      queue.removeAllListeners(`task_failed`)
-      queue.removeAllListeners(`drain`)
-      queue.removeAllListeners(`task_finish`)
+      queue.off(`task_failed`, taskFailedCallback)
+      queue.off(`drain`, drainCallback)
+      if (taskFinishCallback) {
+        queue.off(`task_finish`, taskFinishCallback)
+      }
       queue = null
```

Comment:
Fair warning; last time I tried to change this it caused a bunch of weird side effects and we ended up reverting it.